File: solved/two-sum.cc
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> m(nums);
        sort(m.begin(), m.end());
        vector<int> r;
        for (size_t i = 0; i < nums.size() && r.empty(); i++) {
            auto v = target - nums[i];
            auto p = lower_bound(m.begin(), m.end(), v);
            if (v != *p) {
                continue;
            }
            for (auto j = 0; j < nums.size(); j++) {
                if (nums[j] == v && j != i) {
                    r.push_back(i);
                    r.push_back(j);
                    break;
                }
            }
        }
        return r;
    }
};

File: solved/two-sum.rs
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        use std::cmp::Ordering;
        let mut m = nums.iter().enumerate().collect::<Vec<(usize, &i32)>>();
        m.sort_by(|a, b| a.1.cmp(&b.1));
        for (i, n) in nums.iter().enumerate() {
            match m.binary_search_by(|x| x.1.cmp(&(target - *n)).then(if x.0 == i { Ordering::Less } else { Ordering::Equal })) {
                Ok(p) => return vec![i as i32, m[p].0 as i32],
                _ => {}
            }
        }
        vec![]
    }
}
